File: core/src/main/java/me/devsnox/planetsystem/core/holder/data/PlanetDataImpl.java
package me.devsnox.planetsystem.core.holder.data;

import com.boydti.fawe.FaweAPI;
import com.boydti.fawe.util.EditSessionBuilder;
import com.sk89q.worldedit.EditSession;
import com.sk89q.worldedit.MaxChangedBlocksException;
import com.sk89q.worldedit.blocks.BaseBlock;
import com.sk89q.worldedit.regions.CuboidRegion;
import me.devsnox.dynamicminecraftnetwork.api.DynamicNetworkAPI;
import me.devsnox.dynamicminecraftnetwork.api.DynamicNetworkFactory;
import me.devsnox.planetsystem.api.handler.GridHandler;
import me.devsnox.planetsystem.api.holder.Holder;
import me.devsnox.planetsystem.api.holder.data.PlanetData;
import me.devsnox.planetsystem.api.planet.LoadedPlanet;
import me.devsnox.planetsystem.api.planet.Planet;
import me.devsnox.planetsystem.core.database.DatabasePlanet;
import org.bukkit.Material;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.function.Consumer;

public final class PlanetDataImpl implements PlanetData {

    private final Holder holder;
    private final DynamicNetworkAPI dynamicNetworkAPI;
    private final Set<LoadedPlanet> loadedPlanets;

    public PlanetDataImpl(Holder holder) {
        this.holder = holder;
        this.dynamicNetworkAPI = DynamicNetworkFactory.dynamicNetworkAPI;
        this.loadedPlanets = new HashSet<>();
    }

    @Override
    public void load(UUID owner, Consumer<LoadedPlanet> request) {
        Planet planet = this.holder.getDatabaseHandler().getPlanet(owner).toPlanet();
        planet.load(request);
    }

    @Override
    public void save(UUID owner) {
        System.out.println();
        System.out.println();
        System.out.println("Save Schem" + owner);
        System.out.println();
        System.out.println();
        LoadedPlanet loadedPlanet = this.holder.getPlanetData().getLoadedPlanetByOwner(owner);
        System.out.println(loadedPlanet);
        if (loadedPlanet == null) return;
        DatabasePlanet databasePlanet = DatabasePlanet.by(loadedPlanet);
        System.out.println(databasePlanet);

        this.dynamicNetworkAPI.saveSchematic(databasePlanet.getUuid(), loadedPlanet.getSchematic());
        this.holder.getDatabaseHandler().savePlanet(databasePlanet);

    }

    @Override
    public void unload(UUID owner) {
        this.save(owner);
        LoadedPlanet planet = this.holder.getPlayerData().getPlayer(owner).getPlanet();

        GridHandler grid = this.holder.getGridHandler();
        grid.removeEntry(grid.getId(planet.getMiddle()));

        CuboidRegion cuboidRegion = new CuboidRegion(FaweAPI.getWorld(Holder.Impl.holder.getWorld().getName()), planet.getOuter().getMin().toWEVector(), planet.getOuter().getMax().toWEVector());

        EditSession editSession = new EditSessionBuilder(FaweAPI.getWorld(Holder.Impl.holder.getWorld().getName())).fastmode(true).build();

        try {
            editSession.setBlocks(cuboidRegion, new BaseBlock(Material.AIR.getId()));
        } catch (MaxChangedBlocksException e) {
            e.printStackTrace();
        }

        editSession.flushQueue();

        this.loadedPlanets.remove(planet);
    }

    @Override
    public Set<LoadedPlanet> getLoadedPlanets() {
        return this.loadedPlanets;
    }

}

File: core/src/main/kotlin/de/astride/planetsystem/core/holder/data/PlanetDataImpl.kt
package de.astride.planetsystem.core.holder.data

import com.boydti.fawe.util.EditSessionBuilder
import com.sk89q.worldedit.blocks.BaseBlock
import com.sk89q.worldedit.regions.CuboidRegion
import de.astride.planetsystem.api.functions.toWEWorld
import de.astride.planetsystem.api.holder.Holder
import de.astride.planetsystem.api.holder.data.PlanetData
import de.astride.planetsystem.api.inline.Owner
import de.astride.planetsystem.api.planet.LoadedPlanet
import de.astride.planetsystem.core.database.DatabasePlanet
import de.astride.planetsystem.core.functions.toPlanet
import me.devsnox.dynamicminecraftnetwork.api.DynamicNetworkAPI
import me.devsnox.dynamicminecraftnetwork.api.DynamicNetworkFactory
import org.bukkit.Material
import org.bukkit.entity.Entity
import org.bukkit.entity.Player


//TODO: improve
class PlanetDataImpl(private val holder: Holder) : PlanetData {

    private val dynamicNetworkAPI: DynamicNetworkAPI get() = DynamicNetworkFactory.dynamicNetworkAPI
    override val loadedPlanets: MutableSet<LoadedPlanet> = mutableSetOf()

    override fun load(owner: Owner, request: (LoadedPlanet) -> Unit) =
        holder.databaseHandler.getPlanet(owner).toPlanet().load(request)

    //TODO: Move to LoadedPlanet
    override fun save(owner: Owner) {
        val loadedPlanet = holder.planetData.find(owner) ?: return
        val databasePlanet = DatabasePlanet.by(loadedPlanet)
        dynamicNetworkAPI.saveSchematic(databasePlanet.uuid, loadedPlanet.schematic)
        holder.databaseHandler.savePlanet(databasePlanet)
    }

    //TODO: Move to LoadedPlanet
    override fun unload(owner: Owner) {
        save(owner)

        val planet = holder.playerData.find(owner).planet ?: return
        val distance = (planet.atmosphere.size * 2).toDouble()
        planet.middle.world.getNearbyEntities(planet.middle, distance, distance, distance)

        arrayOf<Entity>().forEach {
            if (it is Player)
                it.teleport(holder.planetData.find(Owner(it.uniqueId)).middle ?: return@forEach)
        }

        holder.gridHandler.removeEntry(holder.gridHandler.getId(planet.middle))

        @Suppress("DEPRECATION")
        EditSessionBuilder(holder.world.toWEWorld()).fastmode(true).build().apply {
            val cuboidRegion = CuboidRegion(this.world, planet.outer.min.toWEVector(), planet.outer.max.toWEVector())
            setBlocks(cuboidRegion, BaseBlock(Material.AIR.id))
            flushQueue()
        }

        loadedPlanets -= planet
    }

}
